lib/vm/src/backends/singlepass.rs

```diff
@@ -34,11 +34,19 @@ pub fn backend() -> Backend {
 }
 
 pub fn set_gas(instance: &mut Instance, limit: u64) {
-    let used = GAS_LIMIT - limit;
+    let used = if limit >= GAS_LIMIT {
+        GAS_LIMIT
+    } else {
+        GAS_LIMIT - limit
+    };
```

Comment:
I'd only go in the special case when the default path `GAS_LIMIT - limit` is invalid, i.e. `limit > GAS_LIMIT`.

And then in the special case, I think `used` must be `0` to express the reset to all available gas.